### PR TITLE
Improved deals & retrieval record tracking & new indices and records API & support verified deals

### PR DESCRIPTION
**Pull request by jsign**
This PR includes new features and improvements to Powergate. 

Lots of things happened here. To improve visibility and review having a mental map of changes, I'll describe the big-picture of changes in different sections.

Most of these changes are motivated by [this Textile PR]([https://github.com/textileio/textile/pull/494](https://github.com/textileio/textile/pull/494))

---

We add verified-deals support in *StorageConfigs*. A new attribute *VerifiedDeal* indicates that new deals will be marked as *verified* in the *deal proposal*.  Of course, the wallet address making the deal should be a verified-client and have enough data-cap to create the verified-deal.

When a verified-deal deal is done, all existing miner-selectors use the *verified-price* from miners instead of the usual *price* for unverified data. The semantics of *MaxPrice* still hold, being verified or unverified, it will block prices that are above *MaxPrice*.

---

A new group of APIs was created to expose indices data for miners. 
These APIs are:

- `GetMiners`: which allows getting a list of all miner-addresses (and no more information) that satisfy some filters (e.g: have power >0 in the network).
- `GetMinersInfo`: Provided with a list of miner-addresses, it will return all known information from indices from these miners, such as ask-price, verified-ask price, relative power, sector-size, min/max piece size, etc.

If an external application wants to update information about miners, it can:

1. Call `GetMiners` to get a list of miner-address of all known miners (this output won't be big since only their address is included).
2. It can batch miners calling `GetMinersInfo` to avoid making big queries. Say, fetching in batches of 30 or 40 miners.

Also, my intention is to let `GetMinersInfo` to include information from the miners index, ask index, or any other internal index. I think this is good to simply have in a single API call consolidated information about the miner. Doing more granular APIs for the different indices doesn't sound like something needed, and also explodes the API surface.

---

A new group of APIs was created to expose deals record fetching, as to understand made deals and retrievals executed by Powergate.

The idea behind these APIs is that an application can fetch incrementally new records that are created or updated with new information:

- `GetUpdatedRetrievalRecordsSince(ctx context.Context, sinceNano int64, limit int)`
- `GetUpdatedStorageDealRecordsSince(ctx context.Context, sinceNano int64, limit int)`

With these APIs, the application can fetch which records changed since some last known `sinceNano` time, and get the updated information.  It can also have some safety-bounds using `limit` to avoid very big outputs if there's too much data to be discovered. If that's the case, it can continue asking for the next set of updated records in a batching mode (`sinceNano` + `limit`).

The way this is implemented is every time a record is created or touched, it's included in an updated-index. 

This change needed to create a migration to allow existing records to be discovered by these new APIs. If no migration is done, querying for the old records would result in zero results. To allow all (old and new) records to be discovered by the same new system, I created a migration that sets the update-index modified timestamp to the existing records' timestamp. This allows you not to have *special cases* where you can't rely on the updated-index or having to create "GetAllRecords" style APIs to get old-things or similar.

---

I took advantage and refactored the deals module in mulitiple files, and simplified the `Watch` API.

Also, the `store` of records now it's in its own package with clearly exported methods. Also, the API of this store was simplified.

---

Before this PR, if a deal failed, it's the corresponding deal-storage record (`Pending: true` ) was deleted. That's not the case now. A new field `Failed` is included, and if the deal fails (or timeouts), it would be set to `true`.

This allows to not lose information about attempted deals that failed. Any attempt to make a deal will be recorded.

---

Retrieval records now also include their `ID` as an attribute. This also needed some work as a migration since existing records didn't have an `ID` value.

The main argument for doing this is because when clients receive retrieval records since their values don't have an ID, it is tough to merge that data with existing data the client has fetched before. Saying it differently, if you don't have an ID... how you "join" to update?

Of course, we could ask the client to do the MD5 of the concatenated 4 values we use internally... but we are not evil with our clients. :)

---

The Ask index now also includes the Verified Ask Price. Before, we only stored the general storage price, but miners also advertise the price if you're making a verified deal.

### CODE REVIEW

File: proto/powergate/user/v1/user.proto
Comment (asutula) [inline proto/powergate/user/v1/user.proto:514]: This one I might push a bit to change before merging. I've convinced myself that sending timestamp data in Protobuf as int isn't ideal. In order to use it properly, the consumer has to know about or make assumptions about the server implementation. This is probably nano seconds, but is it mill? Is it just seconds? Depending on the answer the client needs to process data differently. Some languages handle milliseconds nicely, while others deal with nanoseconds nicely. Anyway, long story short, I think its better to use `import "google/protobuf/timestamp.proto";` and `google.protobuf.Timestamp` to make this most flexible and deterministic on the client side. 

I think there are still times where sending some integer representation of time, paging apis being one of them. But in that case, the integer should be sent as part of the response purely as a token that is sent back with the next request, no meaning other than that, no need to parse or interpret the integer.

**Review by asutula — APPROVED** [review]
This all looks technically very very good. Some great new features an improvements. I left a handful of comments and questions, but only one think to really consider before merging. Other thing I noticed is there's no updates to the cli, just wondering if that was an omission or coming in the future?

File: api/client/admin/options.go
Comment (jsign) [reply api/client/admin/options.go:23]: Uhm, not sure about that since this way allows to set default values. For example, in this case, the default value is `true`. If the API receives a pointer to a struct, then we can't control the default value of options.

File: deals/module/store/store.go
Comment (jsign) [reply deals/module/store/store.go:70]: I think you did it that way to optimize querying by status. 
Now that the Store API is a bit unified, let's wait a bit more to justify merging them. 
Maybe that's not needed.

File: deals/module/store/store.go
Comment (jsign) [reply deals/module/store/store.go:231]: Whenever we can dedicate time to adding app metrics and having good dashboards, we can detect when queries start taking more time than reasonable. At that point, we can think about pruning or doing something.

I think this kind of metrics are important since we're walking in the dark here about how bad this might be.
I guess we eventually need to push this as a prioritization.

File: migration/testdata/v1_Trackstore.post
Comment (jsign) [reply migration/testdata/v1_Trackstore.post:1]: The problem isn't much the system, but much of the migration logic per se.

To let it be fully backward compatible, on every migration we should always use copied structures of the original version. If we use structs directly from the current version, that's faster but not future-compatible.. since as we keep changing powergate, old migrations still reference the "current" structs (not really totally-frozen ones).

We can start doing that, and this problem would disappear. Or we could not do that, and start deprecating old migrations.

File: proto/powergate/user/v1/user.proto
Comment (jsign) [reply proto/powergate/user/v1/user.proto:514]: Yep, sounds good to me. If that datatype is compatible with clients in different languages, sounds like a better option. I guess that's the case, so I'll change it.

**Comment by jsign** [comment]
> Other thing I noticed is there's no updates to the cli, just wondering if that was an omission or coming in the future?
I kept up to date the existing commands with some new flag that I added. Regarding new APIs, yes, they don't have CLI commands and that won't happen in this PR. 

At the same time of doing this PR, I was also working in the `textile` one; and the scope of the epic was too big to consider doing a CLI in a mindful way.

I think your point is part of a bigger topic. At many places, we call "having a CLI command" just printing the raw proto as a JSON output. At some other places, we experimented using other libraries that are more interactive or similar. I could add those CLI commands easily, but I'm not totally convinced that's calling it a day in UX.

In summary, I'd say that this falls into the "coming in the future", but I feel that we should decide if the CLI should have some minimum bar greater than printing a JSON. Not strong thoughts now, but I feel we went back and forth a bit before on that before.

File: proto/powergate/user/v1/user.proto
Comment (jsign) [reply proto/powergate/user/v1/user.proto:514]: Extra point about this: I'll change the new time-related fields to this new better type. The `Time` field I guess you created it as `int64` before this knowing about his, so I'd keep that untouched since that would be a breaking change and this PR is backward -compatible.

We should remember changing all the timestamps to this new type when landing some PR that already justifies breaking the APIs (i.e: v3)

File: api/client/admin/options.go
Comment (asutula) [reply api/client/admin/options.go:23]: Ah didn't realize the default was `true`. Yea, better this way then.

Comment (asutula) [reply api/client/admin/options.go:23]: Well, maybe I take that back. You can just create a default gRPC request with the value set to `true`. If some provided option overrides that, then it is set again. Anyway like I said, not something that has to change, just interesting discussion about patterns.

Comment (jsign) [reply api/client/admin/options.go:23]: Sure, it's worth the discussion :). Yeah, the good thing about default values at the gRPC level is that other clients (say, JS or Python) will have default value too; as it doesn't depend on the client-implementer as here.

File: api/server/admin/indices.go
Comment (jsign) [inline api/server/admin/indices.go:47]: Oh,  also switching prices to string in the API.
We still need to change things internally to avoid using uint64 too, to basically avoided the same problem.
Created https://github.com/textileio/powergate/issues/776 to track.

File: migration/migration4.go
Comment (asutula) [reply migration/migration4.go:124]: Makes sense. Tricky stuff!

**Review by asutula — APPROVED** [review]

**Pull request merged**
